genie-web/src/main/java/com/netflix/genie/web/services/impl/LocalJobKillServiceImpl.java

```diff
@@ -108,6 +118,14 @@ public void killJob(
         @NotBlank(message = "No id entered. Unable to kill job.") final String id,
         @NotBlank(message = "No reason provided.") final String reason
     ) throws GenieException {
+
+        //TODO possible optimization. Have a single projection for getting hostname and v4 job and pass isV4.
+        //That can save one db call
+        if (isV4Job(id)) {
```

Comment:
You may want to put this after the logic that determines if a job is already finished or not below

Reply:
@tgianos I looked at the code and here are my thoughts. I am not clear about how the statuses are being updated today for v4 jobs by the agent. Atleast when a job is killed they are not updated. For e.g Is an exit code being populated for v4 jobs? Its being checked in this block of code 
```final JobExecution jobExecution = this.jobSearchService.getJobExecution(id);
            if (jobExecution.getExitCode().isPresent()) {
                // Job is already finished one way or another
                return;
            } ```
So its not clear to me where to fit the handling of v4 kill accept for right at the beginning. When I had an initial discussion with @mprimi we wanted the initial implementation to be a way (preferably without continuous polling) to convey to a agent that `kill yourself`. Having the handling of v4 jobs right in the beginning, as it is now, serves that purpose. So, we can have this component ready but their is no need for agent to register for kill notification until we have the entire kill functionality(status update, acknowledgement from the agent that its going to kill itself, actual kill functionality on the agent side) figured out. I have also added the code to throw an exception if a kill request comes from a v4 job and there is no registered callback. In case some one makes a kill request there will be an exception thrown clearly communicating that the job cannot be killed. 
This way we should be able to move forward while figuring out other details associated with kill.